R/read-write.R

```diff
@@ -0,0 +1,465 @@
+    list.files(csv_directory),
+    c("___info_file_dm.csv", "___pk_file_dm.csv", "___fk_file_dm.csv", "___coltypes_file_dm.csv")
+  )
+  # sort table_files according to column def_base$table
+  table_files_sorted <- table_files[order(match(table_files, paste0(table_names, ".csv")))]
+
+  table_tibble <- map2(
+    file.path(csv_directory, table_files_sorted),
+    col_class_sorted,
+    ~ readr::read_csv(file = .x, col_types = .y)
+  ) %>%
+    set_names(table_names) %>%
+    enframe(name = "table", value = "data") %>%
+    # empty subset is necessary because
+    # https://www.tidyverse.org/blog/2018/12/readr-1-3-1/#tibble-subclass
+    mutate(data = map(data, `[`))
+
+  make_dm(def_base, table_tibble, pk_info, fk_info)
+}
+
+#' @inheritParams dm_write_csv
+#' @param zip_file_path
+#' For `dm_write_zip`: The file path to the `zip`-file to write.
+#'
+#' For `dm_read_zip`: The file path to the `zip`-file to read the `dm` from.
+#' @param overwrite Logical, default: `FALSE`. In case the file already exists, should it be overwritten?
+#' @details `dm_write_zip()`: write a `dm` to a `zip`-file containing a collection of `csv`-files.
+#' @rdname dm-read-write
+#' @export
+dm_write_zip <- function(dm, zip_file_path, overwrite = FALSE) {
+  if (file.exists(zip_file_path)) {
+    if (overwrite) {
+      message(glue::glue("Overwriting file {tick(zip_file_path)}."))
+    } else{
+      abort_file_exists(zip_file_path)
+    }
+  }
+
+  csv_directory <- file.path(tempdir(), basename(tempfile(pattern = "dm_zip_")))
+  dm_write_csv_impl(dm, csv_directory, zip = TRUE)
+
+  withr::defer(try(unlink(csv_directory, recursive = TRUE)))
+
+  csv_files <- list.files(csv_directory)
+  # compress the file ("-j" junks the path to the file)
+
+  zip(
+    zipfile = zip_file_path,
+    files = file.path(csv_directory, csv_files),
+    flags = "-j -q"
+  )
+  message(glue::glue("Written `dm` as zip-file {tick(zip_file_path)}."))
+  invisible(dm)
+}
+
+#' @details `dm_read_zip()`: read a `dm` from a `zip`-file created using `dm_write_zip()`.
+#' @rdname dm-read-write
+#' @export
+dm_read_zip <- function(zip_file_path) {
+  # FIXME: nicer way for randomized path-name?
+  unzip_directory <- file.path(tempdir(), basename(tempfile(pattern = "dm_unzip_")))
+  withr::defer(unlink(unzip_directory, recursive = TRUE))
+  utils::unzip(zip_file_path, exdir = unzip_directory)
+
+  dm_read_csv(unzip_directory)
+}
+
+#' @inheritParams dm_write_zip
+#' @param xlsx_file_path
+#' For `dm_write_xlsx()`: The file path to the `xlsx`-file to write.
+#'
+#' For `dm_read_xlsx()`: The file path to the `xlsx`-file to read the `dm` from.
+#' @details `dm_write_xlsx()`: write a `dm` to an `xlsx`-file containing several tables defining the `dm`.
+#' @rdname dm-read-write
+#' @export
+dm_write_xlsx <- function(dm, xlsx_file_path, overwrite = FALSE) {
+
+  if (file.exists(xlsx_file_path)) {
+    if (overwrite) {
+      message(glue::glue("Overwriting file {tick(xlsx_file_path)}."))
```

Comment:
How can we mute this message?

Reply:
Is it necessary to be able to mute this? i.e., how often will users recreate/overwrite files? (honestly not sure)
We could add a `quiet` argument